Handle peer addition/removal in a right way

hebasto opened the pull request:
This PR makes `PeerTableModel` handle a peer addition/removal in a right way. See:
- https://doc.qt.io/qt-5/model-view-programming.html#inserting-and-removing-rows
- https://doc.qt.io/qt-5/model-view-programming.html#resizable-models

Fixes #160.

Fixes #191.

hebasto commented:
Rebased 734dd5f95168886e12dc2f07dd91902b1242ffc3 -> d325799acdd8d010d5e4b2f77d9cc67d76d4ad28 ([pr164.01](https://github.com/hebasto/gui/commits/pr164.01) -> [pr164.02](https://github.com/hebasto/gui/commits/pr164.02)) due to the conflict with #162.

hebasto commented:
Will be rebased after any progress in #161 :)

hebasto commented:
Rebased d325799acdd8d010d5e4b2f77d9cc67d76d4ad28 -> f73d340abeedf7e583fe9b2a8b3685a745a9199e ([pr164.02](https://github.com/hebasto/gui/commits/pr164.02) -> [pr164.03](https://github.com/hebasto/gui/commits/pr164.03)) due to the conflict with #161.

promag commented:
Why is this based on #18?

hebasto commented:
> Why is this based on #18?

It fixes selection behavior and handling in general.

jarolrod reviewed (commented):
tACK f73d340abeedf7e583fe9b2a8b3685a745a9199e

Confirming that the PR fixes the issue laid out in #160 on macOS 11.1 with Qt 5.15.2. 

On `master`: If I select a peer, and a peer above it is disconnected, then the peer I had selected becomes unselected. Below are screenshots exhibiting this behavior.

**Master:** I have selected `Peer ID 32`
![peer-is-selected](https://user-images.githubusercontent.com/23396902/104642534-092e6880-5679-11eb-8bcf-935ca2a9df7e.png)

**Master:** `Peer ID 32` is no longer selected because `Peer ID 29` is above it and disconnected
![peer-becomes-deselected](https://user-images.githubusercontent.com/23396902/104642828-6c1fff80-5679-11eb-8b9a-8f3351c9038b.png)

on this `PR`: This behavior no longer occurs. If I select a peer, it remains selected even if a peer above it is disconnected.

**PR:** `Peer ID 14` is selected
![pr-peer-selected](https://user-images.githubusercontent.com/23396902/104643035-b903d600-5679-11eb-8217-9e93a848b90e.png)

**PR:** `Peer ID 0` has disconnected. `Peer ID 14` remains selected.
![pr-peer-remains-selected](https://user-images.githubusercontent.com/23396902/104643106-dcc71c00-5679-11eb-8c43-b47166759c99.png)

hebasto commented:
Rebased f73d340abeedf7e583fe9b2a8b3685a745a9199e -> 32ce58cb18acc6eb3ac3a2c650c38b7bfdb6ad8b ([pr164.03](https://github.com/hebasto/gui/commits/pr164.03) -> [pr164.04](https://github.com/hebasto/gui/commits/pr164.04)) due to the conflict with #180.

hebasto commented:
Rebased 32ce58cb18acc6eb3ac3a2c650c38b7bfdb6ad8b -> a3d21d3dc3d9166a60f050352ea0eff2a8c2140b ([pr164.04](https://github.com/hebasto/gui/commits/pr164.04) -> [pr164.05](https://github.com/hebasto/gui/commits/pr164.05)) due to the conflict with #179.

jarolrod reviewed (commented):
re-ACK a3d21d3dc3d9166a60f050352ea0eff2a8c2140b, tested on macOS 11.1 Qt 5.15.2 after rebase

hebasto commented:
Rebased a3d21d3dc3d9166a60f050352ea0eff2a8c2140b -> fa95a2205e8c88b23e786a382e55f4534e7733ac ([pr164.05](https://github.com/hebasto/gui/commits/pr164.05) -> [pr164.06](https://github.com/hebasto/gui/commits/pr164.06)) due to the conflict with https://github.com/bitcoin/bitcoin/pull/21015.

hebasto commented:
Rebased fa95a2205e8c88b23e786a382e55f4534e7733ac -> b7fe2b1691330166b478f0ced2dd1b6a8835729b ([pr164.06](https://github.com/hebasto/gui/commits/pr164.06) -> [pr164.07](https://github.com/hebasto/gui/commits/pr164.07)) due to the conflict with #166..

jarolrod reviewed (commented):
re-ack b7fe2b1691330166b478f0ced2dd1b6a8835729b, tested on macOS 11.2 Qt 5.15.2 after rebase

Talkless commented on src/qt/clientmodel.h:
```diff
@@ -54,6 +55,7 @@ class ClientModel : public QObject
     interfaces::Node& node() const { return m_node; }
     OptionsModel *getOptionsModel();
     PeerTableModel *getPeerTableModel();
+    PeerTableSortProxy* peerTableSortProxy();
```
nit: could be noexcept, since it just returns pointer. Though since other similar getters here does not use `noexcept`, adding that one kinda seems out of scope.

hebasto replied:
Though since other similar getters here does not use `noexcept`, adding that one kinda seems out of scope.

Going to leave it as is.

promag replied:
FWIW I've suggested that this should not be "global" but rather instanced for each view (currently one in console dialog).

Talkless commented on src/qt/peertablemodel.cpp:
```diff
@@ -223,35 +131,55 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
+    decltype(m_peers_data) new_peers_data;
+    new_peers_data.reserve(nodes_stats.size());
+    for (const auto& node_stats : nodes_stats) {
+        CNodeCombinedStats stats;
```
These stats has quite a few members, and they all gonna be overwritten after construction. Since it's loop, consider initializing in this wat:
```
const CNodeCombinedStats stats{std::get<0>(node_stats), std::get<2>(node_stats), std::get<1>(node_stats)};
```

 It's most dissapointing that `QList` does not have `emplace()` here...

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-797463817).

Talkless commented on src/qt/peertablemodel.h:
```diff
@@ -74,22 +62,25 @@ class PeerTableModel : public QAbstractTableModel
+
 private:
+    //! Internal peer data structure.
+    QList<CNodeCombinedStats> m_peers_data{};
```
nit: `{}` seems redundant, it's not fundamental type like int, pointer or bool to need this kind of explicit default initialization, constructor will be invoked regardless.

hebasto replied:
Yes, it is redundant. Leaving as is to be explicit.

Talkless commented on src/qt/peertablemodel.h:
```diff
@@ -74,22 +62,25 @@ class PeerTableModel : public QAbstractTableModel
-    QVariant data(const QModelIndex &index, int role) const override;
-    QVariant headerData(int section, Qt::Orientation orientation, int role) const override;
-    QModelIndex index(int row, int column, const QModelIndex &parent) const override;
+    int rowCount(const QModelIndex& parent = QModelIndex()) const override;
```
nit: if one likes C++11 toys, one could write `= {}` instead of ` = QModelIndex()`.

hebasto replied:
As stated in the commit description, the goal is to follow Qt [docs](https://doc.qt.io/qt-5/qabstracttablemodel.html#index).

Talkless commented on src/qt/peertablesortproxy.cpp:
```diff
@@ -0,0 +1,43 @@
+#include <QVariant>
+
+PeerTableSortProxy::PeerTableSortProxy(QObject* parent)
+    : QSortFilterProxyModel(parent)
```
nit: since you use `{}` in `m_peer_table_sort_proxy{nullptr};`, maybe worth using `{}` in `QSortFilterProxyModel{parent}` too (meaning in the new code)?

hebasto replied:
As stated in the commit description, the goal is to follow Qt [docs](https://doc.qt.io/qt-5/qabstracttablemodel.html#index).

hebasto replied:
This commit belongs to the #18 on which this one is based.

Talkless commented on src/qt/rpcconsole.cpp:
```diff
@@ -9,13 +9,14 @@
 #include <qt/forms/ui_debugwindow.h>
 
+#include <chainparams.h>
+#include <interfaces/node.h>
```
On My Machine™, it seems only `#include <interfaces/node.h>` is needed?

hebasto replied:
That commit only adds `#include <qt/peertablesortproxy.h>`, and other `include`s are sorted only. I'd leave the need of the `interfaces/node.h` header out of this pr scope.

Talkless replied:
Sorry, missed the fact that it's a sort.

Talkless reviewed (changes requested):
Concept ACK, right fix with nice cleanup.

hebasto commented:
Updated b7fe2b1691330166b478f0ced2dd1b6a8835729b -> 7bb0dc97ab08ebcba84b54031874c0648367d756 ([pr164.07](https://github.com/hebasto/gui/commits/pr164.07) -> [pr164.08](https://github.com/hebasto/gui/commits/pr164.08), [diff](https://github.com/hebasto/gui/compare/pr164.07..pr164.08)):

- addressed @Talkless' [comment](https://github.com/bitcoin-core/gui/pull/164#discussion_r592667897)

Talkless commented on src/qt/peertablemodel.cpp:
```diff
@@ -223,35 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
+    m_node.getNodesStats(nodes_stats);
+    decltype(m_peers_data) new_peers_data;
+    new_peers_data.reserve(nodes_stats.size());
+    for (const auto& node_stats : nodes_stats) {
```
Ugh, my suggestion was not good enought, you can actually move values out of that local `nodes_stats`:
```
    for (auto&& node_stats : nodes_stats) {
        const CNodeCombinedStats stats{std::get<0>(std::move(node_stats)), 
                                       std::get<2>(std::move(node_stats)),
                                       std::get<1>(std::move(node_stats))};
```
(note loop variable is no longer cost reference!)

Or even with an algorithm (also moves, at least it does on local test), if one would care about "No raw loops" mantra :) 
```
    std::transform(nodes_stats.begin(), nodes_stats.end(), std::back_inserter(new_peers_data), [](auto&& i) {
        return CNodeCombinedStats{std::get<0>(std::move(i)), std::get<2>(std::move(i)), std::get<1>(std::move(i))};
    });

```
Though moving `std::transform` seems fishy, not sure if it's "legal" to move source... But gcc 10.2.1 does that.

hebasto replied:
> Ugh, my suggestion was not good enought...

It seems like a kind of premature optimization, no?

Talkless replied:
Maybe, probably. Since there will be mostly on up to ~128 nodes by default, "nobody" would care if there are some extra copies I guess, and `std::transform` variant do look particularly "noisy" (i.e. "optimization" not worth the effort/readability cost).

hebasto commented:
Rebased 7bb0dc97ab08ebcba84b54031874c0648367d756 -> ec3075d7342b65bdd72e34f5705f6cdc8a948b67 ([pr164.08](https://github.com/hebasto/gui/commits/pr164.08) -> [pr164.09](https://github.com/hebasto/gui/commits/pr164.09)) due to the merging of #18.

Talkless reviewed (approved):
tACK ec3075d7342b65bdd72e34f5705f6cdc8a948b67, tested on Debian Sid with Qt 5.15.2. 

Tested using two bitcoind's connected to PR's bitcoin-qt in regtest mode. Did some reconnects, and so far table seems to work fine.

hebasto commented:
Rebased ec3075d7342b65bdd72e34f5705f6cdc8a948b67 -> e102c90e41fb82e534c155bd7cfb9d0e9852c092 ([pr164.09](https://github.com/hebasto/gui/commits/pr164.09) -> [pr164.10](https://github.com/hebasto/gui/commits/pr164.10)) due to the conflict with #311.

promag reviewed (commented):
Tested ACK e102c90e41fb82e534c155bd7cfb9d0e9852c092 on macos.

jonatack commented:
Concept ACK. If this fixes what I think it does, it would be good to see it in v22 as a bug fix.

jarolrod commented on src/qt/peertablemodel.cpp:
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
+        new_peers_data.append(stats);
+    }
+
+    // Handle a new peer addition or an old peer in a right way. See:
```
nit, or something similar
```suggestion
    // Handle new peer addition or removal as suggested in Qt Docs. See:
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

jarolrod commented on src/qt/peertablemodel.cpp:
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
+    // Handle a new peer addition or an old peer in a right way. See:
+    // - https://doc.qt.io/qt-5/model-view-programming.html#inserting-and-removing-rows
+    // - https://doc.qt.io/qt-5/model-view-programming.html#resizable-models
+    // We use the fact that interfaces::Node::getNodesStats
```
```suggestion
    // We take advantage of the fact that the std::vector returned by  interfaces::Node::getNodesStats
```

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

jarolrod commented on src/qt/peertablemodel.cpp:
```diff
@@ -172,19 +131,52 @@ Qt::ItemFlags PeerTableModel::flags(const QModelIndex &index) const
+    // - https://doc.qt.io/qt-5/model-view-programming.html#inserting-and-removing-rows
+    // - https://doc.qt.io/qt-5/model-view-programming.html#resizable-models
+    // We use the fact that interfaces::Node::getNodesStats
+    // returns std::vector naturally sorted by nodeid.
```
```suggestion
    // is sorted by nodeid.
```

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/164#issuecomment-855996526).

jarolrod reviewed (commented):
ACK e102c90e41fb82e534c155bd7cfb9d0e9852c092

Tested functionality on `Arch Linux Qt 5.15.2`, `macOS 11.3 Qt 5.15.2`, and cross-compile from Linux to `Windows 10`.

## Notes on Commits:
- [x] efb7e5aa962d4a4047061996bbb50b6da4592cbc
  - code review ACK cross-checked with Qt docs for [QAbstractTableModel](https://doc.qt.io/qt-5/qabstracttablemodel.html) and [QAbstractItemModel](https://doc.qt.io/qt-5/qabstractitemmodel.html)
- [x] 1b66f6e556631a1a2d89aefba70a79894bd14fcd
  - code review ACK, the `PeerTablePriv` class would no longer be needed.
- [x] e102c90e41fb82e534c155bd7cfb9d0e9852c092
  - code review ACK, patch looks correct in its implementation in following recommendations from Qt docs
  - some nits () 
## Screenshots:

The below screenshots show that this PR will close the mutated version of #160 (see, https://github.com/bitcoin-core/gui/issues/160#issuecomment-855515389). While I am not including videos showing that this PR also fixes #191, I can confirm that it does on all three tested platforms.

### Arch Linux

#### master

| Select Peer  6   | Disconnect Peer 0, Peer 7 is now selected ❌ |
| --------------- | --------------------------------------------- |
| ![master0start](https://user-images.githubusercontent.com/23396902/120961948-cc006b00-c72c-11eb-8c62-510df6b612b8.png) | ![master-disconnect](https://user-images.githubusercontent.com/23396902/120961968-d589d300-c72c-11eb-821b-fd4e9e274b68.png) |

#### PR

| Select Peer  9   | Disconnect Peer 1, Peer 9 remains selected ✅  |
| --------------- | ----------------------------------------------- |
| ![linux-pr-start](https://user-images.githubusercontent.com/23396902/120962058-fe11cd00-c72c-11eb-9066-100eab8ead56.png) | ![linux-pr-disconnect](https://user-images.githubusercontent.com/23396902/120962066-01a55400-c72d-11eb-93a6-eb92ddcb8984.png) |

### macOS 11.3

#### master

| Select Peer  6   | Disconnect Peer 0, Peer 7 is now selected ❌ |
| --------------- | --------------------------------------------- |
| ![Screen Shot 2021-06-06 at 9 17 11 PM](https://user-images.githubusercontent.com/23396902/120947119-be3aed80-c70c-11eb-99e1-ad9a739df5be.png) | ![Screen Shot 2021-06-06 at 9 17 48 PM](https://user-images.githubusercontent.com/23396902/120947129-c430ce80-c70c-11eb-9c5a-8b4f32a07d7f.png) |

#### PR
| Select Peer  8   | Disconnect Peer 0, Peer 8 remains selected ✅ |
| --------------- | ---------------------------------------------- |
| ![macOS-pr-start](https://user-images.githubusercontent.com/23396902/120961800-7e83fe00-c72c-11eb-8ad1-7cdb31ce42ba.png) | ![macOS-pr-disconnect](https://user-images.githubusercontent.com/23396902/120961867-9d829000-c72c-11eb-8aa3-64bb3f7d78ef.png) |

### Windows 10

Note: Difference between UI on master and PR is because master includes ab86ac7739b27f2c45ae72ce4c4ebecf324b3e90

#### master

| Select Peer  4   | Disconnect Peer 0, Peer 6 is now selected ❌ |
| --------------- | --------------------------------------------- |
| ![windows-master-start](https://user-images.githubusercontent.com/23396902/120962261-5e087380-c72d-11eb-8c43-7b56d1c4e7f7.png) | ![windows-master-disconnect](https://user-images.githubusercontent.com/23396902/120962268-6496eb00-c72d-11eb-9aa5-b9295264986b.png) |

#### PR

| Select Peer  5  | Disconnect Peer 1, Peer 5 remains selected ✅ |
| --------------- | ---------------------------------------------- |
| ![windows-start-pr](https://user-images.githubusercontent.com/23396902/120962495-d838f800-c72d-11eb-83d4-5a2a93431282.png) | ![windows-disconnect-pr](https://user-images.githubusercontent.com/23396902/120962519-e2f38d00-c72d-11eb-871c-8103ee39f06d.png) |

hebasto commented:
Updated e102c90e41fb82e534c155bd7cfb9d0e9852c092 -> ecbd91153875c8cdd5b92b840afc116f65e457fb ([pr164.10](https://github.com/hebasto/gui/commits/pr164.10) -> [pr164.11](https://github.com/hebasto/gui/commits/pr164.11), [diff](https://github.com/hebasto/gui/compare/pr164.10..pr164.11)):

- addressed recent @jarolrod's comments

promag reviewed (commented):
reACK ecbd91153875c8cdd5b92b840afc116f65e457fb just improvements to the comment since last review.

jarolrod commented:
re-ACK ecbd911

Only changes since last review are implementation of comment nits
